### PR TITLE
Feature/manage sales

### PR DESCRIPTION
**Pull request by JosMigue**
@TheUrbina20 Could you review this code pls?

### CODE REVIEW

File: app/Http/Controllers/SaleController.php [hunk 1]
```diff
@@ -4,95 +4,94 @@
...
+            $number = Sale::count();
+            $user = Auth::user();
+            $serial = "V".date("Y").date("m").date("d").strval(($number+rand(1,100))+rand(1,1000));
+            $sale = ['user_id'=>$user->id,'serial_number'=>$serial];
+            $completeSale = array_merge($request->validated(),$sale);
+            try {
+                $sale = Sale::create($completeSale);
+                $total = $this->saveSaleDetail($sale->id);
+                $sale->total = $total;
+                $sale->save();
```
Comment (TheUrbina20) [inline app/Http/Controllers/SaleController.php:48]: You are making three db requests. It should be only one.

File: app/Http/Controllers/SaleDetailController.php [hunk 2]
```diff
@@ -2,84 +2,75 @@
...
-     * @return \Illuminate\Http\Response
-     */
-    public function create()
+    public function store(SaleDetailRequest $request)
     {
-        $products = Product::select('id','name','price')->get();
-        return view('sale.create', compact('paymentmethod',$paymentmethod, 'paymentway', $paymentway, 'products', $products));
-    }
+        $product = Product::where('id',$request->validated('product_id'))->first();
+        $products = session('products');
```
Comment (TheUrbina20) [inline app/Http/Controllers/SaleDetailController.php:18]: You don't need this.

File: app/Http/Controllers/SaleDetailController.php [hunk 3]
```diff
@@ -2,84 +2,75 @@
...
-        
-    }
 
-    /**
-     * Show the form for creating a new resource.
-     *
-     * @return \Illuminate\Http\Response
-     */
-    public function create()
+    public function store(SaleDetailRequest $request)
```
Comment (TheUrbina20) [inline app/Http/Controllers/SaleDetailController.php:15]: I don't this this function storing a record.

File: app/Http/Controllers/SaleDetailController.php [hunk 4]
```diff
@@ -2,84 +2,75 @@
...
-    {
-        
-    }
 
-    /**
-     * Show the form for creating a new resource.
-     *
-     * @return \Illuminate\Http\Response
-     */
-    public function create()
```
Comment (TheUrbina20) [inline app/Http/Controllers/SaleDetailController.php:24]: What you probably need to do is request for a sale details creation.
This method can return you a html with the field you need need to insert in your form.

File: app/Http/Controllers/SaleDetailController.php [hunk 5]
```diff
@@ -2,84 +2,75 @@
...
-    /**
-     * Show the form for editing the specified resource.
-     *
-     * @param  int  $id
-     * @return \Illuminate\Http\Response
-     */
-    public function edit($id)
-    {
-        //
+    private function checkIfProductIsAlreadyAdded($products, $productIdAdded){
```
Comment (TheUrbina20) [inline app/Http/Controllers/SaleDetailController.php:43]: You are using the response as a 'yes'/'no' so, a better name can be `IsProductAlreadyAdded`

File: app/Http/Controllers/SaleDetailController.php [hunk 6]
```diff
@@ -2,84 +2,75 @@
...
+        $isProductAdded = -1;
+        foreach ($products as $index => $product) {
```
Comment (TheUrbina20) [inline app/Http/Controllers/SaleDetailController.php:45]: You can map the array or even better...
https://www.php.net/manual/en/function.array-filter.php

File: app/Http/Controllers/SaleDetailController.php [hunk 7]
```diff
@@ -2,84 +2,75 @@
...
-    /**
-     * Remove the specified resource from storage.
-     *
-     * @param  int  $id
-     * @return \Illuminate\Http\Response
-     */
     public function destroy($id)
     {
-        //
+        $products = session('products');
```
Comment (TheUrbina20) [inline app/Http/Controllers/SaleDetailController.php:61]: Since you will remove the whole session stuffs, this probably will become useless

File: public/js/manageItemsTable.js [hunk 8]
```diff
@@ -0,0 +1,206 @@
+function deleteProduct(button){
```
Comment (TheUrbina20) [inline public/js/manageItemsTable.js:1]: Are you seriously using 6 spaces for indentation?

File: public/js/manageItemsTable.js [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+		showCancelButton: true,
+		confirmButtonColor: '#3085d6',
+		cancelButtonColor: '#d33',
+		confirmButtonText: '¡Sí, borrar!'
+	}).then((result) => {
+		if (result.value) {
+			const productId = button.value;
+			$.ajax({
+				headers: {
+					'X-CSRF-TOKEN': $('meta[name="csrf-token"]').attr('content')
```
Comment (TheUrbina20) [inline public/js/manageItemsTable.js:16]: You need to create an ajax utils, instead of add this header.

File: public/js/manageItemsTable.js [hunk 10]
```diff
@@ -0,0 +1,206 @@
+function deleteProduct(button){
+	event.preventDefault();
+	Swal.fire({
```
Comment (TheUrbina20) [inline public/js/manageItemsTable.js:3]: I see that depending of the message type(for example a waring) you define `icon`, `confirmButtonColor`, `CancelButtonColor` exactly the same. Please create an util file, so you can code something like:
`Swal.warning('¿Estas seguro?', 'Esta acción no se puede corregir', 'Sí, borrar')`

Doesn't it looks clearer?
DYRY

File: public/js/manageItemsTable.js [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+					title: 'Oops...',
+					text: `Algo salió mal, intente más tarde ${xhr.status}`
+				})
+			}
+			});
+		}
+	});
+}
+
+function deleteSale(button){
```
Comment (TheUrbina20) [inline public/js/manageItemsTable.js:44]: Why?

File: public/js/manageItemsTable.js [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+			success: (response) => {
+ 				Swal.fire({
+					position: 'center',
+					icon: 'success',
+					title: '¡Venta borrada correctamente!',
+					showConfirmButton: false,
+					timer: 1200
+				});
+				setTimeout(() => {
+					location.reload();
```
Comment (TheUrbina20) [inline public/js/manageItemsTable.js:72]: Why?

Comment (TheUrbina20) [reply public/js/manageItemsTable.js:72]: It totally makes no sense to request this async if you just wait for response and reload the page.
Why not to just make the request with a form or a link?

File: public/js/manageItemsTable.js [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+				})
+			}
+			});
+		}
+	});
+}
+
+function addRowTable(){
+	event.preventDefault();
+    const select = document.getElementById('product-select');
```
Comment (TheUrbina20) [inline public/js/manageItemsTable.js:89]: Indentation here

File: public/js/manageItemsTable.js [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+	$('#table-empty-message').hide('fast');
+	$.ajax({
+		headers: {
+			'X-CSRF-TOKEN': $('meta[name="csrf-token"]').attr('content')
+	},
+		type: 'POST',
+		url: '/salesdetails',
+		data: {product_id : select.value},
+	success: (response) => {
+		table.insertRow(1).innerHTML = (`<tr><td>${selectedOption.text}</td><td>$${productPrice}</td><td><button id="btn-delete-table-item" class="btn btn-danger" value="${select.value}" data-price="${productPrice}" onclick="deleteRowTable(this)"><span class="material-icons">delete</span></button></td></tr>`);
```
Comment (TheUrbina20) [inline public/js/manageItemsTable.js:102]: In this case your response should be the html you building here.
You can avoid building html here, but when needed I think it always better to build with more readable structure. One line it really long an difficult to understand what you are interring in DOM.

File: public/js/manageItemsTable.js [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+				$('#table-empty-message').show('slow');
+			}
+		}else{		 
+			Swal.fire(
+				'¡Algo no me gusta por aquí...!, ',
+				'Recargando página...',
+				'question'
+			);
+			setTimeout(() => {
+				location.reload();
```
Comment (TheUrbina20) [inline public/js/manageItemsTable.js:139]: Same question as above...why?
Jaja just use a link or form

File: public/js/manageItemsTable.js [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+		})
+	}
+	});
+
+
+}
+
+function plusDetail(productPrice){
+	const totalInput = document.getElementById("total-sale");
+	if(Number(totalInput.value) == 0){
```
Comment (TheUrbina20) [inline public/js/manageItemsTable.js:157]: This 6 lineas are not exactly the same as just:
`totalInput.value = (Number(totalInput.value) + productPrice);`

File: public/js/manageItemsTable.js [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+	}else{
+		productPrice+=Number(totalInput.value);
+		totalInput.value = productPrice;
+	}
+}
+
+function substractDetail(button){
+	const productPrice =  Number(button.getAttribute('data-price'));
+	const totalInput = document.getElementById("total-sale");
+	let substract = Number(totalInput.value) - productPrice;
```
Comment (TheUrbina20) [inline public/js/manageItemsTable.js:168]: Is not the same as:
`totalInput.value = Number(totalInput.value) - productPrice;`

File: public/js/manageItemsTable.js [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+			'X-CSRF-TOKEN': $('meta[name="csrf-token"]').attr('content')
+	},
+		type: 'get',
+		url: `/salesdetails/${saleId}`,
+	success: (response) => {
+		alert(response);
+		response.forEach(element => {
+			console.log(element);
+		});
+/* 		if(response){
```
Comment (TheUrbina20) [inline public/js/manageItemsTable.js:185]: If you don't need this, then remove it.

File: public/js/manageItemsTable.js [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+	$.ajax({
+		headers: {
+			'X-CSRF-TOKEN': $('meta[name="csrf-token"]').attr('content')
+	},
+		type: 'get',
+		url: `/salesdetails/${saleId}`,
+	success: (response) => {
+		alert(response);
+		response.forEach(element => {
+			console.log(element);
```
Comment (TheUrbina20) [inline public/js/manageItemsTable.js:183]: I want to believe you learned we should not use logs out of debug purposes and you just forgot to remove it.

File: resources/views/sale/create.blade.php [hunk 20]
```diff
@@ -31,35 +76,47 @@
...
 							<option value="{{$key}}">{{$value}}</option> 
 						@endforeach
 					</select>
 				</div>
-			</div>
-			<div class="text-right">
-				<button class="btn btn-success footer-form-btn" type="submit" >{{__('Add')}}</button>
-			</div>
-		</form>
+				<input type="hidden" name="user_id" id="user_id" value="{{Auth::user()->id}}">
```
Comment (TheUrbina20) [inline resources/views/sale/create.blade.php:93]: I want to believe this id is useless for backend.

File: resources/views/saledetail/show.blade.php [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+  <div class="container">
+    <div class="panel">
+      <div class="panel-heading">
+        <h3>{{__('Sale detail')}}</h3>
+        <a class="btn btn-primary" href="{{route('sales.index')}}">{{__('Get Back')}}</a>
+      </div>
+    </div>
+  </div>
+  <div class="table-responsive container">
+		<table class="table table-striped table-bordered">
```
Comment (TheUrbina20) [inline resources/views/saledetail/show.blade.php:18]: What happened with indentation?

File: app/Http/Controllers/SaleController.php [hunk 22]
```diff
@@ -4,95 +4,94 @@
...
-    public function destroy(Sale $sale)
+
+    public function destroy($sale)
     {
-        //
+        $sale = Sale::findOrFail($sale);
+        if($sale->delete()){
+            return json_encode(array("Success"=>true));
+        }else{
+            return json_encode(array("Success"=>false));
```
Comment (TheUrbina20) [inline app/Http/Controllers/SaleController.php:94]: Does this return a http status out of 200?

File: app/Http/Controllers/SaleDetailController.php (hunk 6 above)
Comment (JosMigue) [reply app/Http/Controllers/SaleDetailController.php]: I tried to use that function, but it only work before, because I used to use only data array where temporally I had my products, but now I'm using array of objects (model instances) from SaleDetail model.

File: app/Http/Controllers/SaleDetailController.php (hunk 7 above)
Comment (JosMigue) [reply app/Http/Controllers/SaleDetailController.php]: The session's stuffs is gone

File: public/js/manageItemsTable.js (hunk 19 above)
Comment (JosMigue) [reply public/js/manageItemsTable.js:183]: You right, I forgot it. sorry

File: app/Http/Controllers/SaleController.php [hunk 23]
```diff
@@ -4,95 +4,86 @@
...
-     * @param  \App\Sale  $sale
-     * @return \Illuminate\Http\Response
-     */
-    public function show(Sale $sale)
-    {
-        //
-    }
+  public function create()
+  {
+    session(['products' => null,]);
```
Comment (TheUrbina20) [inline app/Http/Controllers/SaleController.php:29]: Do you need this?

File: app/Http/Controllers/SaleController.php [hunk 24]
```diff
@@ -4,95 +4,86 @@
...
-     * @return \Illuminate\Http\Response
-     */
-    public function show(Sale $sale)
-    {
-        //
-    }
+  public function create()
+  {
+    session(['products' => null,]); 
+    $paymentmethods = Sale::getEnum('paymentmethod');
```
Comment (TheUrbina20) [inline app/Http/Controllers/SaleController.php:30]: Why you don't use came case?

File: app/Http/Controllers/SaleController.php [hunk 25]
```diff
@@ -4,95 +4,86 @@
...
-    public function destroy(Sale $sale)
-    {
-        //
+  private function isProductAlreadyAdded($product, $details){
+    if($details == null){
+      return -1;
+    }else{
+      foreach ($details as $index => $detail) {
+        if($detail->product_id == $product){
+            return $index;
```
Comment (TheUrbina20) [inline app/Http/Controllers/SaleController.php:73]: Based on the name of your function, you should return a bolean.
Probable you need to create another function for this.

File: app/Http/Controllers/SaleDetailController.php [hunk 26]
```diff
@@ -2,84 +2,80 @@
...
-    /**
-     * Store a newly created resource in storage.
-     *
-     * @param  \Illuminate\Http\Request  $request
-     * @return \Illuminate\Http\Response
-     */
-    public function store(Request $request)
-    {
-        //
+    private function createNewSaleDetail($product){
```
Comment (TheUrbina20) [inline app/Http/Controllers/SaleDetailController.php:20]: User Laravel to build this from an html file instead.

File: app/Http/Controllers/SaleDetailController.php [hunk 27]
```diff
@@ -2,84 +2,80 @@
...
-    /**
-     * Show the form for editing the specified resource.
-     *
-     * @param  int  $id
-     * @return \Illuminate\Http\Response
-     */
-    public function edit($id)
-    {
-        //
+    private function checkIfProductIsAlreadyAdded($products, $productIdAdded){
```
Comment (TheUrbina20) [inline app/Http/Controllers/SaleDetailController.php:50]: Same Boolean logic here as descriptor above.

File: app/Http/Controllers/SaleDetailController.php [hunk 28]
```diff
@@ -2,84 +2,80 @@
...
-     *
-     * @param  int  $id
-     * @return \Illuminate\Http\Response
-     */
-    public function destroy($id)
-    {
-        //
+    public function destroy($id){
+      $products = session('products');
+      $loopIndex = $this->checkIfProductIsAlreadyAdded($products, $id);
```
Comment (TheUrbina20) [inline app/Http/Controllers/SaleDetailController.php:68]: What does loopIndex means?
Why you still using session vars?

File: app/Sale.php [hunk 29]
```diff
@@ -9,18 +9,20 @@ class Sale extends Model
...
     protected $fillable = [
         'customer_id',
+        'serial_number',
+        'user_id',
         'subtotal',
         'VAT',
         'total',
-        'payment_method',
-        'payment_way'
+        'paymentmethod',
```
Comment (TheUrbina20) [inline app/Sale.php:17]: Why did you remove `_`?

File: public/css/theme/error.css [hunk 30]
```diff
@@ -0,0 +1,100 @@
...
+    margin: 0;
+  }
+  
+  #notfound {
+    position: relative;
+    height: 100vh;
+    background: #f6f6f6;
+  }
+  
+  #notfound .notfound {
```
Comment (TheUrbina20) [inline public/css/theme/error.css:17]: `#notfound .notfoud`?

File: public/css/theme/error.css [hunk 31]
```diff
@@ -0,0 +1,100 @@
...
+    -webkit-box-shadow: 0 15px 15px -10px rgba(0, 0, 0, 0.1);
+            box-shadow: 0 15px 15px -10px rgba(0, 0, 0, 0.1);
+  }
+  
+  .notfound .notfound-404 {
+    position: relative;
+    height: 180px;
+  }
+  
+  .notfound .notfound-404 h1 {
```
Comment (TheUrbina20) [inline public/css/theme/error.css:42]: I think all your `h1` should have same style.
Instead you can use something like `h1.primary`, `h1.primary.is-big`, `h1.secondary.is-blue`...

File: public/css/theme/error.css [hunk 32]
```diff
@@ -0,0 +1,100 @@
...
+    font-family: 'Roboto', sans-serif;
+    position: absolute;
+    left: 50%;
```
Comment (TheUrbina20) [inline public/css/theme/error.css:45]: Are you centering text with this? 🤨

File: app/Sale.php (hunk 29 above)
Comment (JosMigue) [reply app/Sale.php:17]: Cus my enums were named like that!

File: app/Http/Controllers/SaleDetailController.php (hunk 28 above)
Comment (JosMigue) [reply app/Http/Controllers/SaleDetailController.php]: It was removed on last commit

File: app/Sale.php (hunk 29 above)
Comment (TheUrbina20) [reply app/Sale.php:17]: Why you didn't create/update you enums according to?

Comment (JosMigue) [reply app/Sale.php:17]: At beginning I didn't know how enums work, that was the reason why I named them like that, I'll try it in while. I need to take a look on that
